### PR TITLE
JBPM-7513 - Create archetype to generate kjar project

### PR DESCRIPTION
**Pull request by tsurdilo**

### CODE REVIEW

**Comment by tsurdilo** [comment]
*****dont merge yet!

@mswiderski do you mind taking a look please?

File: kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/pom.xml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  <properties>
+    <version.org.kie>${project.version}</version.org.kie>
+    <version.org.junit>4.12</version.org.junit>
+    <version.com.thoughtworks.xstream>1.4.10</version.com.thoughtworks.xstream>
+    <maven.compiler.target>1.8</maven.compiler.target>
+    <maven.compiler.source>1.8</maven.compiler.source>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+  </properties>
+
+  <dependencies>
```
Comment (mswiderski) [inline kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/pom.xml]: could you please remove the dependencies? it should not be there for empty projects and only added when needed.

File: kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/src/main/resources/META-INF/kie-deployment-descriptor.xml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#set( $symbol_pound = '#' )
+#set( $symbol_dollar = '$' )
+#set( $symbol_escape = '\' )
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<deployment-descriptor xsi:schemaLocation="http://www.jboss.org/jbpm deployment-descriptor.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+  <persistence-unit>org.jbpm.domain</persistence-unit>
+  <audit-persistence-unit>org.jbpm.domain</audit-persistence-unit>
+  <audit-mode>JPA</audit-mode>
+  <persistence-mode>JPA</persistence-mode>
+  <runtime-strategy>SINGLETON</runtime-strategy>
```
Comment (mswiderski) [inline kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/src/main/resources/META-INF/kie-deployment-descriptor.xml]: maybe we should default to per process instance strategy... we already had this request for workbench projects too

File: kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/src/main/resources/META-INF/persistence.xml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+#set( $symbol_pound = '#' )
+#set( $symbol_dollar = '$' )
+#set( $symbol_escape = '\' )
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
```
Comment (mswiderski) [inline kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/src/main/resources/META-INF/persistence.xml]: this file should not be added by default

File: kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/src/main/resources/WorkDefinition.wid [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    ],
+    "results" : [
+        "Result" : new ObjectDataType(),
+    ],
+    "displayName" : "REST",
+    "icon" : "defaultservicenodeicon.png"
+  ],
+
+  [
+    "name" : "Milestone",
```
Comment (mswiderski) [inline kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/src/main/resources/WorkDefinition.wid]: assuming these are for base projects, so i guess we don't need milestones and subcase work items here

Comment (mswiderski) [reply kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/src/main/resources/WorkDefinition.wid]: maybe instead add decision and business rule tasks... wdyt?

File: kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/global/customeditors.json [hunk 5]
```diff
@@ -0,0 +1,3 @@
+{ "editors":{
```
Comment (mswiderski) [inline kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/global/customeditors.json]: what is this for?

File: kie-archetypes/kie-kjar-archetype/src/main/resources/WorkDefinitions.wid [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+      "Independent" : new StringDataType(),
+      "DestroyOnAbort" : new StringDataType()
+    ],
+    "results" : [
+        "CaseId" : new StringDataType(),
+    ],
+    "displayName" : "Sub Case",
+    "icon" : "defaultsubcaseicon.png",
+    "category" : "Cases"
+  ]
```
Comment (mswiderski) [inline kie-archetypes/kie-kjar-archetype/src/main/resources/WorkDefinitions.wid]: maybe add decision and business rule task work items here, wdyt?

File: kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/src/main/resources/WorkDefinition.wid (hunk 4 above)
Comment (tsurdilo) [reply kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/src/main/resources/WorkDefinition.wid]: It's only for case project - base project does not include a wid by default. When generating base kjar project this file gets removed.

File: kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/global/customeditors.json (hunk 5 above)
Comment (tsurdilo) [reply kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/global/customeditors.json]: it's for legacy designer. this so people can see an example how to create custom editors for parameters (every parameter in old designer can have a custom editor)

File: kie-archetypes/kie-kjar-archetype/src/main/resources/WorkDefinitions.wid (hunk 6 above)
Comment (tsurdilo) [reply kie-archetypes/kie-kjar-archetype/src/main/resources/WorkDefinitions.wid]: added

**Comment by tsurdilo** [comment]
@mswiderski updated. let me know if there is anything else

File: kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/src/main/resources/WorkDefinition.wid (hunk 4 above)
Comment (mswiderski) [reply kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/src/main/resources/WorkDefinition.wid]: ok, we can leave it like this

File: kie-archetypes/kie-kjar-archetype/src/main/resources/WorkDefinitions.wid (hunk 6 above)
Comment (MarianMacik) [reply kie-archetypes/kie-kjar-archetype/src/main/resources/WorkDefinitions.wid]: Just to clarify - what is this file for? Is it included in the generated kjar? It is not under archetype-resources so probably it is not?

File: kie-archetypes/kie-kjar-archetype/README.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+This Archetype is capable of generating both a "base" and a "case" kjar project.
+This is similar to when creating a new project inside the KIE Workbench where you can
+specify one of these two options when creating a new project.
+
+Generate KJAR Base project
+--------------------
+1. build the kie-kjar-archetype module (mvn clean install)
+2. Change to directory of your choice where you want to build the 
+base workitem handler from this archetype.
```
[the remark this answers is not in the document]
Comment (cristianonicolai) [reply kie-archetypes/kie-kjar-archetype/README.md]: +1

File: kie-archetypes/kie-kjar-archetype/README.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+```
+mvn clean install
+```
+
+Generate KJAR Case project
+---------------------------
+
+1. build the kie-kjar-archetype module (mvn clean install)
+2. Change to directory of your choice where you want to build the 
+base workitem handler from this archetype.
```
Comment (cristianonicolai) [inline kie-archetypes/kie-kjar-archetype/README.md]: +1

**Comment by cristianonicolai** [comment]
@tsurdilo @mswiderski overall looks good, my only comment would be to be more jbpm specific with the naming instead of using _kie_, ex the architype artifactId and the default group id that is generated.

File: kie-archetypes/kie-kjar-archetype/src/main/resources/META-INF/archetype-post-generate.groovy [hunk 9]
```diff
@@ -0,0 +1,47 @@
+println "[Running post-generation script]";
+
+def caseProject = "${caseProject}";
+def moduleDir = new File(request.getOutputDirectory()+"/"+request.getArtifactId());
+
+if( caseProject == "true" ) {
+    println "Updating case project...";
+    def baseProjectDDFile = new File(moduleDir, "src/main/resources/META-INF/kie-deployment-descriptor.xml");
+    boolean baseProjectDDFileDeleted =  baseProjectDDFile.delete();
```
Comment (cristianonicolai) [inline kie-archetypes/kie-kjar-archetype/src/main/resources/META-INF/archetype-post-generate.groovy]: any chances of deleting a descriptor that the user has customized, losing the changes?

**Comment by cristianonicolai** [comment]
@mswiderski yep, I get that is generic but if I want to get started creating a process or case project I would probably be looking for a jbpm archetype. If you use a Maven project wizard from an ide like IntelliJ, you can easily recognize that there is an archetype for creating a work item handler project, as it shows under its gav: org.jbpm:jbpm-workitems-archetype. I think the same experience would be nice for creating a case/process project and that would probably be a org.jbpm gav not kie.

File: kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/pom.xml [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>\${groupId}</groupId>
+  <artifactId>\${artifactId}</artifactId>
+  <version>\${version}</version>
+  <packaging>kjar</packaging>
+  <name>\${artifactId}</name>
+
+  <properties>
+    <version.org.kie>${project.version}</version.org.kie>
```
[the remark this answers is not in the document]
Comment (cristianonicolai) [reply kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/pom.xml]: This is probably related to my other comment, making it more jbpm specific and adding the starting point for adding any further dependency based on jbpm. As we don't want to enforce any dependency, we could even add some common ones (jbpm-core, test? ) as commented out examples.

File: kie-archetypes/kie-kjar-archetype/src/main/resources/WorkDefinitions.wid (hunk 6 above)
Comment (tsurdilo) [reply kie-archetypes/kie-kjar-archetype/src/main/resources/WorkDefinitions.wid]: It will be included in the generated kjar under /src/main/resources

**Comment by tsurdilo** [comment]
@cristianonicolai I think ultimately the workitem archetype will end up here so not only all archetypes are in the same place but worktiems can also be used in decision tables so keeping it all "kie" I think is not a bad. This is where I think things like yeoman has advantage as it allows you to search archetypes by defined search strings / capabilities so the gavs don't much matter

File: kie-archetypes/kie-kjar-archetype/pom.xml [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+      </resource>
+      <resource>
+        <directory>src/main/resources</directory>
+        <filtering>false</filtering>
+        <excludes>
+          <exclude>archetype-resources/pom.xml</exclude>
+          <exclude>archetype-resources/project.repositories</exclude>
+        </excludes>
+      </resource>
+    </resources>
```
[the remark this answers is not in the document]
Comment (MarianMacik) [reply kie-archetypes/kie-kjar-archetype/pom.xml]: Thanks, now I see!

File: kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/pom.xml (hunk 10 above)
[the remark this answers is not in the document]
Comment (tsurdilo) [reply kie-archetypes/kie-kjar-archetype/src/main/resources/archetype-resources/pom.xml]: only reason this property is there cause its needed by the 

plugin>
        <groupId>org.kie</groupId>
        <artifactId>kie-maven-plugin</artifactId>
        <version>\${version.org.kie}</version>
        <extensions>true</extensions>
      </plugin>

File: kie-archetypes/kie-kjar-archetype/src/main/resources/WorkDefinitions.wid (hunk 6 above)
Comment (MarianMacik) [reply kie-archetypes/kie-kjar-archetype/src/main/resources/WorkDefinitions.wid]: So what is the difference between this file called workdefinitions.wid and the file workdefinition.wid which is in src/main/resources/archetype-resources/src/main/resources/WorkDefinition.wid?

Comment (tsurdilo) [reply kie-archetypes/kie-kjar-archetype/src/main/resources/WorkDefinitions.wid]: Opps nice catch! I removed this one. Sorry!

**Comment by tsurdilo** [comment]
@cristianonicolai do you think this is OK to be merged or is there anything else? Thanks!

**Comment by cristianonicolai** [comment]
@tsurdilo thanks for the details, I'm fine to merge as is, technically its all good, to me is just a perception/experience thing that I would like to see more _jbpm_ as the final consumable/brand for the community, as I mentioned, archetype have native integration into ides so we could make it easier to get started by promoting a more intuitive namespace.

**Pull request merged**